src/QsCompiler/Transformations/Monomorphization.cs

```diff
@@ -6,47 +6,187 @@
+    {
+        public static Response Apply(Response current)
+        {
+            // Nothing to change if the current callable is already concrete
+            if (current.typeResolutions == null) return current;
+
+            var filter = new ReplaceTypeParamImplementationsSyntax(
+                new MinorTransformations.ReplaceTypeParams(current.typeResolutions.ToImmutableDictionary(param => param.Item1, param => param.Item2)));
+
+            // Create a new response with the transformed callable
+            return new Response
+            {
+                originalName = current.originalName,
+                typeResolutions = current.typeResolutions,
+                concreteCallable = filter.onCallableImplementation(current.concreteCallable)
+            };
+        }
+
+        public ReplaceTypeParamImplementationsSyntax(MinorTransformations.ReplaceTypeParams scope) : base(scope) { }
+
+        public override ResolvedSignature onSignature(ResolvedSignature s)
+        {
+            // Remove the type parameters from the signature
+            s = new ResolvedSignature(
+                ImmutableArray<QsLocalSymbol>.Empty,
+                s.ArgumentType,
+                s.ReturnType,
+                s.Information
+                );
+            return base.onSignature(s);
+        }
+    }
+
+    #endregion
+
+    #region RewriteCalls
+
+    internal class ReplaceTypeParamCallsSyntax :
+        SyntaxTreeTransformation<ScopeTransformation<ReplaceTypeParamCallsExpression>>
+    {
+        public static Response Apply(Response current, Stack<Request> requests, List<Response> responses)
         {
-            return Apply((IEnumerable<QsNamespace>)namespaces);
+            var filter = new ReplaceTypeParamCallsSyntax(new ScopeTransformation<ReplaceTypeParamCallsExpression>(new ReplaceTypeParamCallsExpression(current, requests, responses)));
+
+            // Create a new response with the transformed callable
+            return new Response
+            {
+                originalName = current.originalName,
+                typeResolutions = current.typeResolutions,
+                concreteCallable = filter.onCallableImplementation(current.concreteCallable)
+            };
         }
 
-        public ResolveGenericsSyntax(ScopeTransformation<ResolveGenericsExpression> scope) : base(scope) { }
+        public ReplaceTypeParamCallsSyntax(ScopeTransformation<ReplaceTypeParamCallsExpression> scope) : base(scope) { }
+
     }
 
-    public class ResolveGenericsExpression :
-        ExpressionTransformation<ExpressionKindTransformation<ResolveGenericsExpression>>
+    internal class ReplaceTypeParamCallsExpression :
+        ExpressionTransformation<ReplaceTypeParamCallsExpressionKind>
     {
-        public ResolveGenericsExpression() : base(ex => new ExpressionKindTransformation<ResolveGenericsExpression>(ex as ResolveGenericsExpression)) { }
+        private Response CurrentResponse;
```

Comment:
Looks like all of these (CurrentResponse, Requests, Responses) could be readonly?

Reply:
The need for this is removed in the other PR: #206 .